### PR TITLE
Fix for: [Admin][Permissions] Granting permission to a shop displays it in "Producers" list as if it was a supplier #9589

### LINKED ISSUE #9589: [Admin][Permissions] Granting permission to a shop displays it in "Producers" list as if it was a supplier
## Description
When Shop A grants permission to Shop B to sell its product, these two things happen:
👍- Shop A will appear in Shop B shopfront page, in the Producers list -> this is OK 
👎- Shop B will appear in Shop A shopfront page, in the Producers list -> this is not OK , as Shop B is a **distributor** for Shop A, and not a **supplier**. 

This has been raised by a shop manager who didn't understand why many producers that were not supplying him products appeared in his shop.

## Expected Behavior
A shop should appear in a shop's "Producers" list only if it supplies this shop, which means only if it has granted the permission to this shop to sell its products.

## Actual Behaviour
When granting a permission from Shop A to Shop B, Shop B appears in Shop A producers list (boutique-1/shop#/producers).
(It only applies when Shop B is a producer).

## Steps to Reproduce
1. Connect to a producer's shop (shop A)
2. Grant permission "add to order cycle" from this producer to another **producer** shop or hub or profile (shop B)
3. Go to shop A url and click on the "Producers" tab (boutique-1/shop#/producers)
4. See that shop B appears 

## Animated Gif/Screenshot
Screen Shot 2022-08-22 at 20 58 36
Screen Shot 2022-08-22 at 21 02 33
Screen Shot 2022-08-22 at 21 08 11

## Workaround
Remove permission from Shop A to Shop B (but then Shop B won't be able to sell products from Shop A).

## Severity
bug-s3: a feature is broken but there is a workaround

### PR DESCRIPTION
**Pull request by ashwini-seshadri**
#### What? Why?

Closes #9589

Fix for: [Admin][Permissions] Granting permission to a shop displays it in "Producers" list as if it was a supplier.

This PR resolves the problem whereby Shop B shows on Shop A's producers list even though Shop B is its distributor when Shop A grants permission to Shop B.

This PR also assures that a shop will only be listed as a "Producer" in a shop's "Producers" list if it supplies to that specific shop, which means only if it has granted that specific shop permission to sell its products.

#### What should we test?

- Connect to a producer's shop (shop A)
- Grant permission "add to order cycle" from this producer to another producer shop (shop B)
- Go to shop A url and click on the "Producers" tab
- See that shop B does not appear
- Other parts of the app are not affected by the change
 

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

#### Dependencies

No

#### Documentation updates

No

Screenshot 2022-10-11 at 9 46 29 PM
Screenshot 2022-10-11 at 9 46 54 PM
Screenshot 2022-10-11 at 9 47 05 PM

### CODE REVIEW

File: app/models/enterprise.rb [hunk 1]
```diff
@@ -259,9 +268,9 @@ def relatives
     ", id, id)
   end
 
-  def plus_relatives_and_oc_producers(order_cycles)
+  def plus_parents_and_order_cycle_producers(order_cycles)
     oc_producer_ids = Exchange.in_order_cycle(order_cycles).incoming.pluck :sender_id
-    Enterprise.not_hidden.is_primary_producer.relatives_of_one_union_others(id, oc_producer_ids | [id])
+    Enterprise.not_hidden.is_primary_producer.parents_of_one_union_others(id, oc_producer_ids | [id])
```
Comment (mkllnk) [inline app/models/enterprise.rb:273]: The old scope is now unused and can be removed.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Ah, great. Now just a bit of cleanup, I think.

File: app/models/enterprise.rb (hunk 1 above)
Comment (ashwini-seshadri) [reply app/models/enterprise.rb:277]: Hi, I have removed the unused scope and pushed the commit!

**Review by mkllnk — APPROVED** [review]
I hope that the specs will pass now.

**Review by jibees — APPROVED** [review]
Thanks a lot!

File: app/models/enterprise.rb (hunk 1 above)
Comment (mkllnk) [reply app/models/enterprise.rb:277]: Oh, I think that we were working on it at the same time and I did the same plus some spec fixes. Sorry to override your changes. It should be all good now though.

**Review by mkllnk — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @ashwini-seshadri,
Just to let you know. I am currently testing your PR systematically because there are so many combinations of permissions. Please be patient and in the meantime feel free to pick up another issue if you like!
Thanks! :blush:

**Comment by drummer83** [comment]
Hi @ashwini-seshadri,
Thank you so much for this pull request. I believe this is a really important one, because it is aiming straight at the heart of OFN which is about short supply chains and their network. These should always be displayed correctly.

I decided to test this pull request systematically to discover further problems beyond this issue. So I created a complete set of permissions and checked if they were applied and displayed correctly. 

In general there are 4 possible identified failures:
**!1!** Producer shop should not be displayed as a place to buy producer profile's or another producer shop's or producer hub's products because by definition a producer shop only sells own products.
**!2!** Producer profile or another producer shop or producer hub should not be displayed as a producer of producer shop's products, because by definition the producer shop only sells own products.
**!3!** B granted permission to add to order cycle to A, but A didn't grant permission  to B, so B shouldn't be listed as a place to buy A's products.
**!4!** B granted permission to add to order cycle to A, but A didn't grant permission to B, so A shouldn't be listed as a producer of B's products.

Here you can see test results **BEFORE** you pull request. I did not do the complete set because it is quite time consuming.
| Permission | Shop/hub modal | Profile modal | Shops -> our producers | Producers -> Shop for ... | Shop -> Producers |
|-|-|-|-|-|-|
| producer profile grants permission to producer profile     | -              | OK            | OK                     | OK                        | -                 |
| producer profile grants permission to producer shop        | OK             | !1! :x:       | !2! :x:                | !1! :x:                   | !2! :x:           |
| producer profile grants permission to producer hub         | OK             | OK            | OK                     | OK                        | OK                |
| producer profile grants permission to non-producer profile | -              | OK            | -                      | OK                        | -                 |
| producer profile grants permission to non-producer hub     | OK             | OK            | OK                     | OK                        | OK                |
| producer shop grants permission to producer profile        | OK             | !1! !3! :x:   | !2! !4! :x:            | !1! !3! :x:               | !2! !4! :x:       | 
| producer shop grants permission to producer shop           | !1! !3! :x:    | -             | !2! !4! :x:            | !1! !3! :x:               | !2! !4! :x:       |
| producer shop grants permission to producer hub            | !1! !3! :x:    | -             | !2! !4! :x:            | !1! !3! :x:               | !2! !4! :x:       |

I then tested the complete set of permissions **AFTER** your pull request:
| Permission                                                     | Shop/hub modal | Profile modal | Shops -> our producers | Producers -> Shop for ... | Shop -> Producers |
|-|-|-|-|-|-|
| producer profile grants permission to producer profile          | -              | OK            | OK                     | OK                        | -                 |
| producer profile grants permission to producer shop             | OK             | !1! :x:       | !2! :x:                | !1! :x:                   | !2! :x:           |
| producer profile grants permission to producer hub              | OK             | OK            | OK                     | OK                        | OK                |
| producer profile grants permission to non-producer profile      | -              | OK            | -                      | OK                        | -                 |
| producer profile grants permission to non-producer hub          | OK             | OK            | OK                     | OK                        | OK                |
| producer shop grants permission to producer profile             | OK             | !1! !3! :x:   | OK                     | !1! !3! :x:               | OK                |
| producer shop grants permission to producer shop                | !1! !3! :x:    | -             | !2! :x:                | !1! !3! :x:               | !2! :x:           |
| producer shop grants permission to producer hub                 | !1! !3! :x:    | -             | OK                     | !1! !3! :x:               | OK                |
| producer shop grants permission to non-producer profile         | OK             | OK            | OK                     | OK                        | OK                |
| producer shop grants permission to non-producer hub             | OK             | -             | OK                     | OK                        | OK                |
| producer hub grants permission to producer profile              | OK             | !3! :x:       | OK                     | !3! :x:                   | OK                |
| producer hub grants permission to producer shop                 | !1! !3! :x:    | -             | !2! :x:                | !1! !3! :x:               | !2! :x:           |
| producer hub grants permission to producer hub                  | !3! :x:        | -             | OK                     | !3! :x:                   | OK                |
| producer hub grants permission to non-producer profile          | OK             | OK            | OK                     | OK                        | OK                |
| producer hub grants permission to non-producer hub              | OK             | -             | OK                     | OK                        | OK                |
| non-producer profile grants permission to producer profile      | -              | OK            | -                      | OK                        | -                 |
| non-producer profile grants permission to producer shop         | OK             | OK            | OK                     | OK                        | OK                |
| non-producer profile grants permission to producer hub          | OK             | OK            | OK                     | OK                        | OK                |
| non-producer profile grants permission to non-producer profile  | -              | OK            | -                      | -                         | -                 |
| non-producer profile grants permission to non-producer hub      | OK             | OK            | OK                     | -                         | OK                |
| non-producer hub grants permission to producer profile          | OK             | !3! :x:       | OK                     | !3! :x:                   | OK                |
| non-producer hub grants permission to producer shop             | OK             | !3! :x:       | OK                     | !3! :x:                   | OK                |
| non-producer hub grants permission to producer hub              | !3! :x:        | -             | OK                     | !3! :x:                   | OK                |
| non-producer hub grants permission to non-producer profile      | OK             | OK            | OK                     | OK                        | OK                |
| non-producer hub grants permission to non-producer hub          | OK             | -             | OK                     | OK                        | OK                |

## Summary
- You can see the that all of the errors !4! are gone after staging your pull request! That's fantastic and a great improvement and exactly what the issue was about.
- You can also see, that there are still quite a few errors even after your pull request, but that was out of scope of the original issue.
- I can't find any problems or regressions with your pull request, I only see great improvement.

This pull request is ready to go! :rocket: :muscle: 
I will open a follow-up issue of this one - maybe you feel like working on that as well.

**Comment by mkllnk** [comment]
Great testing, thank you!

**Pull request merged**